### DISCUSSION TITLE
Graceful Reset of I2C on RPI Pico after Timeout

### DISCUSSION THREAD

**Discussion by jdmorise** [micropython/micropython#12547] [RP2040 / Pico]
I am developing a continuously running temperature monitor with a RPI Pico W, a DS18B20 temperature sensor (OneWire interface) and an SSD1306-based OLED Display (on an I2C bus). The display shows the current time and temperature, and updates every 15 seconds. After a random longer period of time between 10 minutes and several hours, the I2C responds with a timeout, and the RPI needs to be rebooted to continue. Currently, a watchdog timer is used to reboot the device with such kind of an error, but I am looking for a cleaner implementation. 

The error message with the call hierarchy: 
``` 
Traceback (most recent call last):
  File "<stdin>", line 236, in <module>
  File "<stdin>", line 218, in main
  File "<stdin>", line 133, in meas_send_loop
  File "oled.py", line 54, in fb_temp_log
  File "ssd1306.py", line 100, in show
  File "ssd1306.py", line 118, in write_data
OSError: [Errno 110] ETIMEDOUT
```

with the failing function 

```py
def write_data(self, buf):
        self.write_list[1] = buf
        self.i2c.writevto(self.addr, self.write_list)
```

I wonder a decent error handling and a graceful reset of the I2C bus could be implemented. 
The Timeout error itself can be handled with 
```py 
  try: 
      [...]
  except OSError:
       print('OLED Timed out')    
```
However, the reset of the I2C bus was not possible so far, the program could not be resumed and the display stayed black. 
In the error handler, I tried to initialise the I2C again with the same settings: 
`i2c = I2C(0, scl=Pin(9), sda=Pin(8), freq=400000)       # Init I2C using pins GP8 & GP9 (default I2C0 pins)
`
The `I2C.deinit()` function is not supported on RPI, this also does not work. Could anybody explain how to properly reset the I2C bus? 

Further hints on getting the I2C more stable and prevent the error are also highly appreciated. 

Further information on board and software : 
RPI Pico W with micropyton release `RPI_PICO_W-20230426-v1.20.0.uf2`.

**Comment by davefes** [comment]
How are you getting the traceback message?  rshell, mpremote, an IDE?

**Reply by jdmorise** [reply]
I use Virtual Studio Code as IDE with the Pico-W-Go extension. VS Code offers the REPL prompt in the console. From the REPL, I start the main program and am able to get prints and error messages.

**Comment by rkompass** [comment]
I found this [application note](https://www.analog.com/media/en/technical-documentation/application-notes/54305147357414AN686_0.pdf).

While the master controls the SCL line, a slave may continue pulling the sda line down, as it is open drain, thus making the i2c communication impossible.

The recommended solution is.

> 1) Master tries to assert a Logic 1 on the SDA line
> 2) Master still sees a Logic 0 and then generates a clock pulse on SCL (1-0-1 transition)
> 3) Master examines SDA. If SDA = 0, go to Step 2; if SDA = 1, go to Step 4
> 4) Generate a STOP condition

You would have to reinit Pins 8 and 9 as digital pins. like `sda = Pin(9); scl = Pin(8)` outside the I2C init call.

For the STOP condition e.g. [here](https://onlinedocs.microchip.com/pr/GUID-04DAA7D3-9FC2-45F2-B757-157190106490-en-US-2/index.html?GUID-93E7ADC6-2D36-4BF8-B022-C59D805E11D7):

> The I2C specification defines a Stop condition as the transition of the SDA line from an active state to an idle state while the SCL line is idle. The master will issue a Stop condition when it has completed its transactions and is ready to release control of the bus, or if a bus time-out occurs.

![I2C_stop](https://github.com/micropython/micropython/assets/90282516/1d4535e1-06d1-46be-9ea6-72c8959385ff)

For T_SU:STO you have to look up the datasheet of the SSD1306, but I suppose there a sufficiently long time of e.g. above 12 ordinary scl clock cycles would suffice.

Afterwards I would start with the reinit of the i2c object and try with an i2c scan first whether the disply is there.

**Reply by rkompass** [reply]
Update: According to the docs we have `i2c.stop()`.

If you succeed with the reset procedure above (please give feedback) I would suggest making an enhancement proposal (as issue). Given how simple i2c is with Micropython (one of its real advantages) this should be made complete.

**Reply by jdmorise** [reply]
Thanks for the detailed reply. I will give the proposals a try over the weekend and will post the outcome.
